Added rendering for ref and local_ref on public_transport=stop_positions

map-per opened the pull request:
Fixes #4662

Changes proposed in this pull request:
- Added rendering for ref and local_ref on public_transport=stop_positions. The track numbers are only shown when train=yes is set
- Track numbers display from zoom 18 on

Test rendering with links to the example places:
(before left, right after)

https://www.openstreetmap.org/node/2470201868
![MHB17](https://user-images.githubusercontent.com/79519062/229798048-a6758b27-2780-4ce1-99b8-b5019ac13d7c.png)
![MHB18](https://user-images.githubusercontent.com/79519062/229798084-84bc1a6b-a5f5-4b3f-8298-a858eb0b4006.png)
![MHB19](https://user-images.githubusercontent.com/79519062/229798105-2a1547f6-b671-4139-af04-9419003dbae5.png)

https://www.openstreetmap.org/node/2465304880
![OST17](https://user-images.githubusercontent.com/79519062/229798225-3624abf6-d004-4f98-957c-3dd06cee96bc.png)
![OST18](https://user-images.githubusercontent.com/79519062/229798176-63baf1c9-a0f8-425d-a89c-754616a3dc42.png)
![OST19](https://user-images.githubusercontent.com/79519062/229798285-554155c0-e053-407b-b8bc-5f57dc12b21b.png)

https://www.openstreetmap.org/node/2476438979
![Pasing18](https://user-images.githubusercontent.com/79519062/229798343-2a310417-f376-4383-b362-420235442dd1.png)
![Pasing19](https://user-images.githubusercontent.com/79519062/229798364-1fb1a9ea-494b-4897-b18c-9eb32cbe3a54.png)
![Pasing20](https://user-images.githubusercontent.com/79519062/229798386-6c38d278-864a-46b9-a2ae-8fbd896a9676.png)

https://www.openstreetmap.org/node/3135325888
![Neu17](https://user-images.githubusercontent.com/79519062/229798486-19faa751-cbfd-4f78-9e32-bc1dc351bcfe.png)
![Neu18](https://user-images.githubusercontent.com/79519062/229798513-3a6d53b2-f01f-453e-8480-f9e5628f03e9.png)
![Neu19](https://user-images.githubusercontent.com/79519062/229798530-7e9d4308-213d-4e17-90b8-784143704e92.png)

https://www.openstreetmap.org/node/260406611
![Freising18](https://user-images.githubusercontent.com/79519062/229798606-b680cb59-087c-4c69-9d20-331879fcdb62.png)

https://www.openstreetmap.org/node/2499357757
![MUC19](https://user-images.githubusercontent.com/79519062/229798644-e36bc63c-ee03-4168-b310-4cb03335477e.png)

https://www.openstreetmap.org/node/3707303125
![Fisch18](https://user-images.githubusercontent.com/79519062/229798693-7bd55106-7d9b-4208-9b3f-23769138e55e.png)

https://www.openstreetmap.org/node/3183841635
![Altona18](https://user-images.githubusercontent.com/79519062/229798804-bb6ea0be-3956-48b6-abf9-1cb84bf0f5be.png)

pnorman commented on project.mml:
```diff
@@ -1411,6 +1411,22 @@ Layer:
+          tags->'local_ref' as local_ref
+        FROM planet_osm_point
+        WHERE (tags->'public_transport' = 'stop_position') AND (tags->'train' = 'yes') AND ((ref IS NOT NULL) OR ((tags->'local_ref') IS NOT NULL))
+        ) AS stop_positions
```
This needs an ordering, otherwise it will be inconsistent between adjancent tiles

map-per replied:
Thanks, I tried some ordering but it didn't change anything. What should I order by?

map-per replied:
Looks like the cut off labels are a general issue of the docker setup and are not caused by this pullrequest (see #4808)

pnorman replied:
Ordering should be by something that makes sense for the priority of the features and the very last thing should be osm_id

map-per replied:
I tried different orderings two weeks ago and in the end ended up with osm_id which should garante a consistent ordering. Today I removed the entier order by section as I don't think that the ordering has anything to do with the rendering artifacts (see #4808 ).

map-per replied:
These are collisions caused by one tile not considering what is rendered in the next tile. Ordering could only contribute to solving collisions between two track numbers, but the observed collisons are between track numbers and other objects.

I don't think that this issue will occur in the productive tile rendering for osm.org as this appears to be a problem fo the docker test setup, but it would be good to double check this of course.

pnorman commented on project.mml:
```diff
@@ -1411,6 +1411,22 @@ Layer:
+      table: |-
+        (SELECT
+          way, 
+          tags->'public_transport' as public_transport,
```
With only one possible value, there's no reason to include this in the output.

map-per replied:
Fixed

pnorman commented on project.mml:
```diff
@@ -1411,6 +1411,22 @@ Layer:
+          ref,
+          tags->'local_ref' as local_ref
+        FROM planet_osm_point
+        WHERE (tags->'public_transport' = 'stop_position') AND (tags->'train' = 'yes') AND ((ref IS NOT NULL) OR ((tags->'local_ref') IS NOT NULL))
```
This needs to follow the [SQL style guidelines](https://github.com/gravitystorm/openstreetmap-carto/blob/master/CONTRIBUTING.md#sql-style-guidelines)

pnorman replied:
Still an issue.

> To check if a tag is in the tags hstore, use tags @> 'foo=>bar', relying on automatic conversion from text to hstore

pnorman reviewed (changes requested):
Not a full review - just a few technical issues, including those related to problems you were experiencing.

map-per commented:
@pnorman Thanks!

imagico commented:
Thanks for the pull request.

This is not a review but i like to mention that as a general principle we do not want to newly introduce interpretation of several tags as synonyms in this style when mappers have not reached a consensus on tagging.  If two tags are consistently used for different things the option to render them identically is there - but i don't see this being the case here.

Beyond that i would like to see input from people experienced in public transport mapping on this.  What is shown here in the examples is the local numbering/labeling of tracks or platforms within train stations.  It would be important to know if the interpreted tagging is commonly used for that in different parts of the world.

I also wonder what the conventions are for the location of `public_transport=stop_position` nodes.  The way this change renders them incentivizes placing them where it is convenient to have the label, usually probably around the middle of the train when it stops.  If mapping conventions suggest a different location (like the front of the train or some other reference point) the mapping incentives we create might clash with that.

pnorman commented on project.mml:
```diff
@@ -1411,6 +1411,23 @@ Layer:
     properties:
       cache-features: true
       minzoom: 12
+  - id: stop-positions
```
Can you name it `train-stop-positions` or something similar to make it clear it's not all stop positions?

pnorman reviewed (changes requested):
There remain SQL style issues and ordering problems raised in previous reviews.

Only one of `ref` or `local_ref` should be used. My preference is `ref`.

The position of this layer relative to others doesn't make much sense. Right now it's lower priority than only place names and stations. This will lead to other features being rendered up to zoom 18, and then train stop positions will come up earlier in the rendering and collide them away. This is sometimes okay, but generally not a good idea because it's confusing to have features disappear like that.

Given the size and zoom, my inclination would be to put these either before or after the *-low-priority layers.

Note: I haven't reviewed the MSS or cartography yet, but the issues raised here are independent of those.

map-per commented:
Thanks for the quick review, I fixed the SQL style, changed the PR to only render ref and not local_ref and moved the priority befor the low-priority layers.

@imagico I'm not an expert for public transprot mapping, but as far as I now placing the `stop_position` in the middle of the train / paltform is conventional. This is also what I overved when taking a look at stop_positions around the world with overpass.

osm-ToniE commented:
@pnorman : "Only one of ref or local_ref should be used. My preference is ref."

The 'ref' should be seen as the global, unique ID for the stop (in the country?) and does not reflect the platform/track of the station.
This is a more generic name for the stop ID. Other, similar tags are 'ref:IFOPT' and 'uic_ref' 

The 'local_ref' is the the platform/track number of the station which should be rendered.

osm-ToniE commented:
@imagico : "I also wonder what the conventions are for the location of public_transport=stop_position nodes."

I agree to the view of @map-per : this is mostly seen as the middle of the train/platform - for trains.

stop_positions for buses are different - mostly where the front door of the bus will be when the bus stops.

imagico commented:
Maybe i need to explain my request more clearly.  I am asking about mapping practice to avoid the need to spend time analyzing tag use of features i am unfamiliar with and this way to facilitate a faster review.  But this only works if i can rely on the answers i get.  I had a look myself now - growing a bit suspicious with the conflation of statements about de facto and ought-to-be tag use in the answers and the result is that while placement of `public_transport=stop_position` roughly in the middle of the train where it stops is indeed widespread in many countries it is also clear that this is not broad consensus world wide and there are quite a few countries where this is distinctly not the case.  A few examples

https://www.openstreetmap.org/node/10027694711
https://www.openstreetmap.org/node/10602804886
https://www.openstreetmap.org/node/4269245658
https://www.openstreetmap.org/node/5836393505
https://www.openstreetmap.org/node/6010790033
https://www.openstreetmap.org/node/1271716495
https://www.openstreetmap.org/node/2367900477
https://www.openstreetmap.org/node/619293646
https://www.openstreetmap.org/node/7070782103
https://www.openstreetmap.org/node/1148313241

So i restate my request: _I would like to see input from people experienced in public transport mapping on this._ That means critical assessments of people knowledgeable in the domain about this change and the assumptions it makes on mapping and tagging practice, not just affirmative statements by people who would like to see this change merged.

As said: This change aims to label the local numbering/labeling of tracks or platforms within train stations.  Therefore it is important to know if and to what extent the assumptions this makes on mapping and tagging practice for that purpose are actually reflecting consensus among mappers in various parts of the world and to what extent and where mapping conventions exists that differ and therefore would clash with this.

It would also be useful to have a review of how other maps with global scope (in particular public transport focused ones) handle the  numbering/labeling of tracks or platforms within train stations.

map-per commented:
@imagico: ToniE is experienced in public transport mapping, he even runs his own public transport quality assurance tool (https://ptna.openstreetmap.de). That's why I asked him to take a look at this.

@osm-ToniE: Thanks for taking a look at this!

OpenGreenStreet commented:
@imagico Is there anything else I can contribute here? (if so, I would be very happy to receive an e-mail in German).

imagico commented:
What is needed from my perspective here is a thorough analysis of the world wide mapping practice.

There are different concepts of numbering (platform/stop/track/route numbers with different scope of application) and there are different methods to record such numbers in OSM.  These seem primarily:

* `ref` on `railway=rail`
* `railway:track_ref` on `railway=rail`
* `ref` on route relations with `route=track`
* `ref` on `railway=platform`
* `local_ref` on `railway=platform`
* `ref` on `railway=platform_edge`
* `ref` on `public_transport=stop_position`
* `local_ref` on `public_transport=stop_position`

In some cases these are further differentiated into simple numbers and semicolon separated lists.  In addition - as mentioned above - the placement practice of `public_transport=stop_position` is a further point that needs consideration.

To decide if we can productively show any of these and how we should possibly do so we would need to know in more detail how these different methods are used to record what kind of numbers and how this practice of use differs globally.  I have only looked at this enough to see that there is substantial variability that we cannot ignore.  But that does not necessarily mean we can't render anything.  But we need reliable information on this to determine viability.

Feel free to provide information in languages other than English if that is easier - we will try to make use of that.  If reading English is an issue please try using online translation services (like [deepl](https://www.deepl.com/)).

OpenGreenStreet commented:
@imagico : Another mapper drew my attention to www.openrailwaymap.org. Does this possibly help with a solution? : https://www.openrailwaymap.org/?lang=de&lat=51.8815775618944&lon=14.063186645507812&zoom=10&style=standard

osm-ToniE commented:
@imagico there seems to be no common sense on that. Even inside the station München Hauptbahnhof, shown in the first comment, it is not consistent.

- on a platform_edge (way) I saw local_ref, the platform itself as the MP it is ref
- [openrailwaymap.org](https://www.openrailwaymap.org/?lang=de&lat=51.8815775618944&lon=14.063186645507812&zoom=10&style=standard) seems to show the value of railway:track_ref on the way
- [Transport Map layer](https://www.openstreetmap.org/way/39688751#map=18/48.14062/11.55725&layers=T) doesn't show rail/tram infrastructure but shows the value of local_ref on (bus) stops (on highway=bus_stop and/or [public_transport=stop_position](https://www.openstreetmap.org/way/39688751#map=19/48.08946/11.64479&layers=T))

I would use **local_ref** on **public_transport=stop_position** as **priority 1**
- this **public_transport=stop_position** is generic  and defined for any public transport vehicle
  - it is clear where to map this: **a node on the way**, not besides the way, not a way or area or MP itself
- **local_ref** is the most specific tag here: preferred to be used here 
  - it is clear that the scope of that id is local
- **ref** is a more generic tag which is used for many other purposes (not preferred to be used here)
  - can be confused using ref:IFOPT (a European standard, unique id stop_id inside Europe)

If **local_ref** is not set, one might consider using **ref** instead?
- not a good id? ref might already be set without having the 'platform'/'stop' in mind

My 2ct: use **local_ref** on **public_transport=stop_position** and propagate/postulate/announce this as being '**the choice**' for Carto

OpenGreenStreet commented:
@osm-ToniE , @imagico :

Thank you for addressing this issue. It would be really helpful if there was a solution.
